### ISSUE TITLE
[API Proposal]: Add a type to represent IP networks based on CIDR notation

### ISSUE THREAD

**Issue by geeknoid** [dotnet/runtime#79946]
### Background and motivation

We've been using a custom IPAddressRange in our project, but this seems like a general-purpose concern. Looking across internal code bases at Microsoft, we've found dozens of similar implementations of this type, reinforcing the notion that this is general-purpose and should be in the framework.

### API Proposal

*Edit by @antonfirsov:* Added implementations for `ISpanFormattable` and `IParsable<IPSubnet>`, harmonized naming of method arguments to match current BCL standards.

```csharp
namespace System.Net;

// Alternative name ideas: IPSubnet, IPAddressSpace
public readonly struct IPNetwork : IEquatable<IPSubnet>, ISpanFormattable, ISpanParsable<IPSubnet>
{
    public IPAddress BaseAddress { get; }
    public byte MaskLength { get; }

    public IPNetwork(IPAddress baseAddress, byte maskLength);

    public bool Contains(IPAddress address);

    // ISpanParsable
    public static IPNetwork Parse(string s, IFormatProvider? provider = null);
    public static bool TryParse(string s, IFormatProvider? provider, out IPNetwork result);
    public static IPNetwork Parse(ReadOnlySpan<char> s, IFormatProvider? provider = null);
    public static bool TryParse(ReadOnlySpan<char> s, IFormatProvider? provider, out IPNetwork result);

    // Parsing - Convenience methods
    public static bool TryParse(string s, out IPNetwork result);
    public static bool TryParse(ReadOnlySpan<char> s, out IPNetwork result);

    // ISpanFormattable
    public string ToString(string? format, IFormatProvider? provider = null);
    public bool TryFormat(Span<char> destination, out int charsWritten, ReadOnlySpan<char> format = default, IFormatProvider? provider = null);

    // Equality
    public bool Equals(IPNetwork value);
    public static bool operator ==(IPNetwork left, IPNetwork right);
    public static bool operator !=(IPNetwork left, IPNetwork right);
}
```

### API Usage

We're using this for parsing and logging, so the parsing and tostring functionality are particularly interesting.

### Alternative Designs

~This type could potentially implement `ISpanFormattable<T>` and `ISpanParsable<T>`, but those interfaces require an IFormatProvider instance which isn't relevant in this context. Still, those interfaces might provide better interop elsewhere in the framework. Note that `IPAddress` doesn't implement these interfaces either.~

**Comment by teo-tsirpanis** [comment]
Should it have a `bool TryFormat(Span<char>, out int)`?

**Comment by NN---** [comment]
When I saw IPAddress range, I assumed a range of address not only by a bitmask.
E.g.
`AddRange(IPAddress start, IPAddress end)`

**Comment by svick** [comment]
This looks like a duplicate of https://github.com/dotnet/runtime/issues/64033 and https://github.com/dotnet/runtime/issues/36428, which were closed as being too niche.

Note that ASP.NET Core does have [the `IPNetwork` class](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.httpoverrides.ipnetwork?view=aspnetcore-7.0), though it's quite limited.

**Comment by sakno** [comment]
[IPNetwork2](https://www.nuget.org/packages/IPNetwork2) perfectly serves to that purpose.

**Comment by geeknoid** [comment]
> This looks like a duplicate of #64033 and #36428, which were closed as being too niche.

Definitely in the eye of the beholder :-) As I said, we've found dozens of implementations of this within Microsoft's various code bases.
 
> Note that ASP.NET Core does have [the `IPNetwork` class](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.httpoverrides.ipnetwork?view=aspnetcore-7.0), though it's quite limited.

Missed that one. From a composition perspective, it's a bit too high in the stack for what we'd like I think. But it does show that the concept is a useful one.

**Comment by geeknoid** [comment]
> Should it have a `bool TryFormat(Span<char>, out int)`?

Definitely could, and it could also implement ISpanFormattable/ISpanParsable. Note that IPAddress could also use these.

**Comment by teo-tsirpanis** [comment]
And how about we call it `IPAddressSubnet`? It more closely matches what it is.

**Comment by geeknoid** [comment]
@mddddb

**Comment by mddddb** [comment]
> And how about we call it `IPAddressSubnet`? It more closely matches what it is.

IMO in order to make this an actual IP address range maybe we should make this more flexible in terms of the First address and the Last address in the range, i.e. cover the cases of custom address ranges that cannot be covered by CIDR notation (subset of addresses in a subnet).

So we can have both
`(IPAddress begin, IPAddress end)` and `(IPAddress baseAddress, byte maskLength)` constructors, and more if needed.
Although I am not sure if we need to have some kind of a property specifying whether the range can be considered valid in CIDR.

**Comment by geeknoid** [comment]
@mddddb Then the MaskLength property would need to be removed or made optional or something, since a range might not in fact be just a mask thing.

**Comment by mddddb** [comment]
@geeknoid Yep, I am thinking either a nullable prop or a method with nullable return type

**Comment by karelz** [comment]
Triage: Prioritizing it for 8.0 as the number of ad-hoc implementations clearly shows demand.
During design we should take into account previous asks pointed out by @svick (thanks!) above: https://github.com/dotnet/runtime/issues/79946#issuecomment-1364504712

**Comment by antonfirsov** [comment]
> IMO in order to make this an actual IP address range maybe we should make this more flexible in terms of the First address and the Last address in the range, i.e. cover the cases of custom address ranges that cannot be covered by CIDR notation 

This doesn't sound like a feature that would be highly used. Instead I would just go with a class name that is less confusing, eg. `IPNetwork` sounds good enough.

**Comment by geeknoid** [comment]
@antonfirsov What's the motivation for not implementing ISpanFormattable?

**Comment by antonfirsov** [comment]
@geeknoid we would also need to implement it on `IPAddress` then, otherwise the API would look inconsistent. I'm in favor keeping this proposal as lean as possible in order to get it approved and done quickly. Implementing `ISpanFormattable` on both `IPAddress` and `IPSubnet/Network` can be a follow-up enhancement.

**Comment by antonfirsov** [comment]
I realized now that since `IPAddress` is mutable, `readonly struct` would not guarantee actual immutability of entire instances. Shall we turn the type into a class instead to avoid the false illusion of an immutable type? Do we expect many allocations from instances?

**Comment by geeknoid** [comment]
readonly struct does not guarantee deep immutability.

We generally prefer structs for things that are small and not extensible. We do have a couple scenarios which maintain a bunch of these things in collections, so having those as structs is helpful there. Not a big deal either way though.

**Comment by ilmax** [comment]
> * The name should be `IPSubnet` or `IPNetwork`. `IPSubnet` is slightly preferred, because it doesn't collide with aspnet's `IPNetwork`.

I'm not sure `IPSubnet` is the best naming option, I found it a bit confusing to include the subnet suffix. To me `IPAddressRange` or `IPAddressSpace` seems like a better name.

**Comment by antonfirsov** [comment]
> To me `IPAddressRange` or `IPAddressSpace` seems like a better name.

The term "Range" usually means an A-B range of countable items in .NET, many people would find it confusing, see discussion above. The term "Space" reminds me of the entire space of IPv4 or IPv6 addresses, not the address of a subnetwork, I'm not aware of frameworks/libraries using this term for this purpose.

**Comment by antonfirsov** [comment]
I will update the proposal according to suggestions in https://github.com/dotnet/runtime/issues/79946#issuecomment-1387499626.

The main open question for me is whether we should create a `class` or a `readonly struct`, see my https://github.com/dotnet/runtime/issues/79946#issuecomment-1387281284. @geeknoid can you comment on that? Any other opinions?

**Comment by MihaZupan** [comment]
> since IPAddress is mutable, readonly struct would not guarantee actual immutability of entire instances.

The main "mutable"-ness you're referring to is the setter on `IPAddress.Address`?

Given that that property is already marked as obsolete (and IMO having the setter in the first place feels like a mistake), I'd be fine exposing this type as a `readonly struct`. That better reflects the intent that it's meant to be read-only. If we want to be extra safe, we can store the `IPAddressRange.BaseAddress` as a `ReadOnlyIPAddress` as part of parsing.

**Comment by ilmax** [comment]
> The term "Range" usually means an A-B range of countable items in .NET, many people would find it confusing

Isn't it what it represents? A range of possible ip addresses starting from a point `IPAddressRange.BaseAddress` up to what the mask allows?

Are you referring to the fact that it's not a collection so it doesn't actually contains anything?

**Comment by antonfirsov** [comment]
@ilmax you can't define an arbitrary range, eg. `192.168.0.253-192.168.0.255` (inclusive) with IP + mask length -> we cannot define an `IPAddressRange(IPAddress start, IPAddress end)` constructor -> the proposed model doesn't match the general concept of a range like it works with other .NET API-s that have the term "Range" in their names.

**Comment by antonfirsov** [comment]
> The main "mutable"-ness you're referring to is the setter on `IPAddress.Address`?

> Given that that property is already marked as obsolete

@MihaZupan we also have [ScopeId](https://learn.microsoft.com/en-us/dotnet/api/system.net.ipaddress.scopeid?view=net-7.0) but that doesn't interfere with masking, so good point, `readonly struct` looks reasonable to me now.

**Comment by geeknoid** [comment]
> The main open question for me is whether we should create a `class` or a `readonly struct`, see my [#79946 (comment)](https://github.com/dotnet/runtime/issues/79946#issuecomment-1387281284). @geeknoid can you comment on that? Any opinions from others?

I would prefer a struct given it's a bit more efficient overall, but would be happy with a class as well.

**Comment by antonfirsov** [comment]
Updated the proposal with according to the outcomes of the discussion and a naming harmonization. If there are no objections against the current shape, I think this can be marked ready for review.

**Comment by antonfirsov** [comment]
@ilmax can you elaborate why do you find the name `IPSubnet` confusing? I want to understand how users think about such a name. At this point `IPNetwork` feels better to me TBH.

**Comment by ilmax** [comment]
@antonfirsov I find it a bit misleading because you can use CIDR notation also to define a network address space, so I would like not to see the word **subnet** in there. I suggested to use `IPAddressSpace` instead because it's more general term and in Azure it's used extensively. 

Having said that, I'm not a networking expert so I shyly state my opinion and I'm ok with whatever name you will end up using as long as we get the type into the bcl.

**Comment by antonfirsov** [comment]
Indeed, that name doesn't work well [in context of Azure](https://learn.microsoft.com/en-us/azure/virtual-network/concepts-and-best-practices), switching the preferred name in the proposal to `IPNetwork` which seems to be a very common name for such a type in other frameworks/libraries.

**Comment by bartonjs** [comment]
* We changed the mask length from `byte` to `int`.
* Since this type doesn't support custom formatting, the ToString/TryFormat/Parse/TryParse methods should be explicit interface implementations and have public exposure without the format/provider arguments.
* Based on IPv6 / CIDR calling it "prefix" vs "mask", we've changed `maskLength` to `prefixLength`

```C#
namespace System.Net;

// Alternative name ideas: IPSubnet, IPAddressSpace
public readonly struct IPNetwork : IEquatable<IPNetwork>, ISpanFormattable, ISpanParsable<IPNetwork>
{
    public IPAddress BaseAddress { get; }
    public int PrefixLength { get; }

    public IPNetwork(IPAddress baseAddress, int prefixLength);

    public bool Contains(IPAddress address);

    // ISpanParsable (explicit)
    static IPNetwork ISpanParsable<IPNetwork>.Parse(string s, IFormatProvider? provider = null);
    static bool ISpanParsable<IPNetwork>.TryParse(string s, IFormatProvider? provider, out IPNetwork result);
    static IPNetwork ISpanParsable<IPNetwork>.Parse(ReadOnlySpan<char> s, IFormatProvider? provider = null);
    static bool ISpanParsable<IPNetwork>.TryParse(ReadOnlySpan<char> s, IFormatProvider? provider, out IPNetwork result);

    // Parsing - Public methods
    public static IPNetwork Parse(string s);
    public static IPNetwork Parse(ReadOnlySpan<char> s);
    public static bool TryParse(string s, out IPNetwork result);
    public static bool TryParse(ReadOnlySpan<char> s, out IPNetwork result);

    // ISpanFormattable (explicit)
    string ISpanFormattable<IPNetwork>.ToString(string? format, IFormatProvider? provider = null);
    bool ISpanFormattable<IPNetwork>.TryFormat(Span<char> destination, out int charsWritten, ReadOnlySpan<char> format = default, IFormatProvider? provider = null);

    // Formatting (public)
    public override string ToString();
    public bool TryFormat(Span<char> destination, out int charsWritten);

    // Equality
    public bool Equals(IPNetwork value);
    public static bool operator ==(IPNetwork left, IPNetwork right);
    public static bool operator !=(IPNetwork left, IPNetwork right);
}
```

**Comment by antonfirsov** [comment]
@geeknoid are you interested to port your implementation to dotnet/runtime? If yes, any ETA?

**Comment by mddddb** [comment]
@antonfirsov @karelz @geeknoid A [draft PR ](https://github.com/dotnet/runtime/pull/82779)has been created, there are couple of questions I left there. There is still some polishing and documentation left to be done, although I am not sure what the preference is on that part - should I continue with the changes? Or will someone take it from here? Anyway, some feedback is appreciated before the PR gets published

**Comment by antonfirsov** [comment]
@mddddb thanks! I will take a look tomorrow. It would be nice if you can react to basic feedback, but I'm happy to take over more complicated stuff, eg. the changes needed if we decide to ship it as a NuGet package.

**Comment by mddddb** [comment]
@antonfirsov I am happy to make the necessary changes, as long as we don't have burning deadlines. This is an interesting exercise and a new experience to me, I just have to juggle this together with other tasks that I have🙂

**Comment by antonfirsov** [comment]
We should consider producing a netstandard2.0-consumable OOB package for this type similarly to #65577 / #79120.
Following that precedent, I propose the package name **Microsoft.Bcl.Net.Primtives**, that should target `netstandard2.0`, `net8.0` (with typeoforward), and probably .NET Framework if that's our preferred practice.

One challenge here is that we cannot implement `IEquatable<IPNetwork>`, `ISpanFormattable` and `ISpanParsable<IPNetwork>` for the old targets, therefore those API-s should be conditional to net8.0+.

```C#
namespace System.Net;

public readonly struct IPNetwork : 
#if NET8_0_OR_GREATER
  IEquatable<IPNetwork>, ISpanFormattable, ISpanParsable<IPNetwork>
#else
  IFormattable
#endif
{
    public IPAddress BaseAddress { get; }
    public int PrefixLength { get; }

    public IPNetwork(IPAddress baseAddress, int prefixLength);

    public bool Contains(IPAddress address);

#if NET8_0_OR_GREATER
    // ISpanParsable (explicit)
    static IPNetwork ISpanParsable<IPNetwork>.Parse(string s, IFormatProvider? provider = null);
    static bool ISpanParsable<IPNetwork>.TryParse(string s, IFormatProvider? provider, out IPNetwork result);
    static IPNetwork ISpanParsable<IPNetwork>.Parse(ReadOnlySpan<char> s, IFormatProvider? provider = null);
    static bool ISpanParsable<IPNetwork>.TryParse(ReadOnlySpan<char> s, IFormatProvider? provider, out IPNetwork result);
#endif

    // Parsing - Public methods
    public static IPNetwork Parse(string s);
    public static IPNetwork Parse(ReadOnlySpan<char> s);
    public static bool TryParse(string s, out IPNetwork result);
    public static bool TryParse(ReadOnlySpan<char> s, out IPNetwork result);

    // IFormattable (explicit)
    string IFormattable<IPNetwork>.ToString(string? format, IFormatProvider? provider = null);

#if NET8_0_OR_GREATER 
    // ISpanFormattable (explicit)
    bool ISpanFormattable<IPNetwork>.TryFormat(Span<char> destination, out int charsWritten, ReadOnlySpan<char> format = default, IFormatProvider? provider = null);
#endif

    // Formatting (public)
    public override string ToString();
    public bool TryFormat(Span<char> destination, out int charsWritten);

    // Equality
    public bool Equals(IPNetwork value);
    public static bool operator ==(IPNetwork left, IPNetwork right);
    public static bool operator !=(IPNetwork left, IPNetwork right);
}
```

**Comment by antonfirsov** [comment]
@scottleyg this is a lightweight API focusing on core functionality - parsing and `Contains`. The API in #36428 seems to be really heavyweight, useful advanced for tools dealing with CIDR.

>  It should be CidrBlock

I actually considered that name, but "IP Network" seemed to be the a more widespread name for types dealing with this concern, eg. go has [`IPNet`](https://pkg.go.dev/net#IPNet), python has [`ip_network`](https://docs.python.org/3/library/ipaddress.html#ipaddress.ip_network), aspnet has (a now to be obsoleted) [`IPNetwork`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.httpoverrides.ipnetwork?view=aspnetcore-7.0). The only example I was able to find that uses the word CIDR in their type names is rust (trait [`Cidr`](https://docs.rs/cidr/latest/cidr/trait.Cidr.html), but then the docs immediately say _"Types implementing Cidr represent IP networks"_. I don't think there is a name that is ideal from all perspectives, and I think `IPNetwork` should be recognizable for most users, thus I find it to be an ok compromise. We are past API-review now anyways, and I'm about to merge the implementing PR.

If you think have some valuable additions to the [approved API](https://github.com/dotnet/runtime/issues/79946#issuecomment-1438985842), which are important enough for common scenarios to be included in the BCL, feel free to open a new API proposal, but bare in mind that we will likely not commit to advanced/rare stuff.

**Comment by antonfirsov** [comment]
We are keeping this open to track downlevel packaging.

**Comment by antonfirsov** [comment]
We have a bunch of packages [prefixed `Microsoft.Bcl`](https://www.nuget.org/packages?q=microsoft.bcl) and recently #79120 created the package `Microsoft.Bcl.Cryptography`. We can introduce a new naming scheme for new packages going forward, but IMO it will create more friction. /cc @terrajobst who's working on a design doc for OOB-ing.

**Comment by danmoseley** [comment]
> We can introduce a new naming scheme for new packages going forward, but IMO it will create more friction.

yeah fair enough I got that feedback offline.

**Comment by karelz** [comment]
Triage: Remaining work is to decide on downlevel packaging - waiting for API Review Board's decision.

**Comment by tarekgh** [comment]
Is this still not decided yet? just trying to track the progress, no pressure.

**Comment by karelz** [comment]
I created separate issue to track the netstandard2.0 decision - #88965

**Comment by karelz** [comment]
Fixed in #82779
